src/System.Slices/src/System/Contract.cs

```diff
@@ -31,17 +31,26 @@ public static void Requires(bool condition)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static void RequiresInRange(int start, int length)
         {
-            if (!(start >= 0 && start < length))
+            // explicit unchecked block is used in order to avoid OverflowException when casting 
+            // negative int to unsigned int when the code is compiled with /checked option
+            unchecked
             {
-                throw new ArgumentOutOfRangeException();
+                // use unsigned int to reduce range check by one ( >= 0 )
+                if ((uint)start >= (uint)length)
```

Comment:
Seems like the helper assumes nonnegative length. It makes sense to use it in checks like index range checking  (since we know the span length is nonnegative by construction). In fact this is exactly the same technique as used by JIT for array index operations.
Perhaps it is worth adding Debug.Assert(length >= 0) ?